File: android/src/main/kotlin/com/example/pl_image_downloader/services/DownloadBridge.kt
package com.example.pl_image_downloader.services

import com.example.pl_image_downloader.models.DownloadCallBack
import com.example.pl_image_downloader.utils.ChannelTag
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch


/**
 * DownloadBridge
 * A bridge to communicate download progress and status updates from native Android to Flutter.
 */
class DownloadBridge(
    flutterEngine: FlutterEngine
) {

    companion object {
        const val TAG = "DownloadBridge"
    }

    /** Coroutine scope for managing asynchronous tasks. */
    val scope: CoroutineScope =  CoroutineScope(Dispatchers.Main.immediate + SupervisorJob())

    private val channel = MethodChannel(
        flutterEngine.dartExecutor.binaryMessenger,
        ChannelTag.SERVICE_CHANNEL
    )


    /** Invokes a progress update on the Flutter side. */
    fun invokeProgress(progress: Int, id: Long) {
        val callBack = DownloadCallBack.Progress(
            value = progress,
            id = id
        )
        scope.launch {
            channel.invokeMethod(ChannelTag.EVENT_BRIDGE, callBack.toMap())
        }
    }


    /** Disposes the coroutine scope to clean up resources. */
    fun disposeScope() {
        scope.cancel()
    }

}

File: ios/Classes/DownloadBridge.swift
//
//  DownloadBridge.swift
//
//  Created by Nguyễn Minh Hưng on 8/5/26.
//



struct DownloadBridge {
    static private let TAG = "DownloadBridge"

    private  var batteryChannel: FlutterMethodChannel?
    
    public init(messenger: FlutterBinaryMessenger) {
        batteryChannel = FlutterMethodChannel(name: ChannelTag.serviceChannel, binaryMessenger: messenger)
    }
    
    func invokeProgress(progress: Int, id: Int64) {
        print("\(DownloadBridge.TAG) invokeProgress: \(progress) id: \(id)")
        
        let callBack = DownloadCallBack.onProgress(value: progress, id: id)
        
        DispatchQueue.main.async {
            self.batteryChannel?.invokeMethod(ChannelTag.eventBridge, arguments: callBack.toMap())
        }
    }
    
    func invokeResult(result: DownloadResult, id: Int64) {
        print("\(DownloadBridge.TAG) invokeResult: id: \(id)")
        
        DispatchQueue.main.async {
            self.batteryChannel?.invokeMethod(ChannelTag.eventBridge, arguments: result.toMap())
        }
    }
    
    func dispose() {
        print("\(DownloadBridge.TAG) dispose")
    }
}
